enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,185 @@
+
+## Summary
+
+This enhancement allows a cluster administrator to upgrade a deployed Windows
+Machine Config Operator (WMCO), and associated Windows node components, to a
+newer version.
+
+## Motivation
+
+The main motivation behind this enhancement is to give the end user
+an easy and non-disruptive way to upgrade to a newer release of the WMCO, and
+upgrade associated Windows node components with it.
+WMCO is not part of the release payload, and is managed with [OLM](https://github.com/operator-framework/operator-lifecycle-manager).
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Enable upgrading WMCO with as little disruption as possible.
+* Enable upgrading on-node Kubernetes components through newer versions of the
+  WMCO, including major version upgrades across OpenShift cluster upgrades, and
+  bug and CVE fixes during the current version.
+
+### Non-Goals
+
+In this enhancement we do not plan to support Windows operating system updates.
+
+## Proposal
+
+To allow for the WMCO to be upgraded in a safe and non-disruptive way, it is
+proposed that each Windows VM configured by previous versions of the WMCO has
+its Machine object deleted, resulting in the drain and deletion of the Windows
+node, and the termination and recreation of the VM. The upgraded WMCO instance
+will then be able to configure VMs that will be created in place of the
+terminated ones.
+
+In order to allow for this to occur with minimal service disruption, it is
+recommended that users have at least two Windows nodes within the cluster.
+
+All Windows Kubernetes component updates will be tied to WMCO releases.
+
+### Justification
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in-place upgrade,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. However, in-place upgrades do not provide any key benefits
+for Windows workloads. For example, Windows has [restrictions regarding storage](https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#storage)
+preventing the projection of volumes from host storage, because of this we do
+not have to worry about preserving node local data.
+With all aspects weighed, the risk of introducing additional issues to provide
+an in-place upgrade is not worth it at this time, however it should be explored
+as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+### Design Details
```

Comment:
Since we're allowing RBAC for machine resources, we should specify what we think we'll need.  We'll need delete and read, I think.  Maybe this is too specific, I don't consider it a blocker.

Anyway, what we definitely need is to call out that this operator has elevated privileges and should only be scheduled onto control-plane hosts.  I'm not sure how this fits into the design with OperatorHub, but it needs to be explicit here.

Reply:
Hmm, the scheduling may be a challenge. @shawn-hurley is this possible with an OLM managed operator?

Reply:
I agree that it would be important for WMCO to run on control-plane hosts. There is complexity in that change as WMCO is an optional component and an investigation should be done to see if doing that is possible. If so, the [WMCO enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/windows-machine-config-operator.md) should be updated to reflect that, as that would be out of scope of this enhancement, which deals with upgrades. As a note, WMCO already has Machine read permissions as seen in the linked enhancement.

Reply:
@sebsoto ACK, lets follow up on that separately.

Reply:
I think since this enhancement deals with adding RBAC that allows deleting machine objects, it's not outside the scope of this enhancement.  If we can't support running WMCO exclusively on control plane or some type of infra host, then we shouldn't proceed with giving the pod those privileges.  For me, this would be a blocker.